solidity-verifiers/templates/nova_cyclefold_decider.askama.sol

```diff
@@ -1,24 +1,162 @@
+        }
+        return limbs;
+    }
+}
+
 /**
  * @author  PSE & 0xPARC
- * @title   NovaDecider contract, for verifying zk-snarks Nova IVC proofs.
- * @dev     This is an askama template. It will feature a snarkjs groth16 and a kzg10 verifier, from which this contract inherits.
- *          WARNING: This contract is not complete nor finished. It lacks checks to ensure that no soundness issues can happen.
- *          Indeed, we know some of the checks that are missing. And we're working on the solution
- *          but for now, it's good enough for testing and benchmarking.
+ * @title   NovaDecider contract, for verifying Nova IVC SNARK proofs.
+ * @dev     This is an askama template which, when templated, features a Groth16 and KZG10 verifiers from which this contract inherits.
  */
 contract NovaDecider is Groth16Verifier, KZG10Verifier {
-    function verifyProof(uint[2] calldata _pA, uint[2][2] calldata _pB, uint[2] calldata _pC, uint[1] calldata _pubSignals, uint256[2] calldata c, uint256[2] calldata pi, uint256 x, uint256 y) public view returns (bool) {
-        bool success_kzg = super.check(c, pi, x, y);
-        require(success_kzg == true, "KZG  Failed");
+    /**
+     * @notice  Computes the linear combination of a and b with r as the coefficient.
+     * @dev     All ops are done mod the BN254 scalar field prime
+     */
+    function rlc(uint256 a, uint256 r, uint256 b) internal pure returns (uint256 result) {
+        assembly {
+            result := addmod(a, mulmod(r, b, BN254_SCALAR_FIELD), BN254_SCALAR_FIELD)
+        }
+    }
+
+    /**
+     * @notice  Verifies a nova cyclefold proof consisting of two KZG proofs and of a groth16 proof.
+     * @dev     The selector of this function is "dynamic", since it depends on `z_len`.
+     */
+    function verifyNovaProof(
+        // inputs are grouped to prevent errors due stack too deep
+        uint256[{{ 1 + z_len * 2 }}] calldata i_z0_zi, // [i, z0, zi] where |z0| == |zi|
+        uint256[4] calldata U_i_cmW_U_i_cmE, // [U_i_cmW[2], U_i_cmE[2]]
+        uint256[3] calldata U_i_u_u_i_u_r, // [U_i_u, u_i_u, r]
+        uint256[4] calldata U_i_x_u_i_cmW, // [U_i_x[2], u_i_cmW[2]]
+        uint256[4] calldata u_i_x_cmT, // [u_i_x[2], cmT[2]]
+        uint256[2] calldata pA, // groth16 
+        uint256[2][2] calldata pB, // groth16
+        uint256[2] calldata pC, // groth16
+        uint256[4] calldata challenge_W_challenge_E_kzg_evals, // [challenge_W, challenge_E, eval_W, eval_E]
+        uint256[2][2] calldata kzg_proof // [proof_W, proof_E]
+    ) public view returns (bool) {
+
+        require(i_z0_zi[0] >= 2, "Folding: the number of folded steps should be at least 2");
+
+        // from gamma_abc_len, we subtract 1. 
+        uint256[{{ public_inputs_len - 1 }}] memory public_inputs; 
+
+        public_inputs[0] = i_z0_zi[0];
+
+        for (uint i = 0; i < {{ z_len * 2 }}; i++) {
+            public_inputs[1 + i] = i_z0_zi[1 + i];
+        }
+
+        {
```

Comment:
Right, the verifier compute the NIFS.V (step 10 [here](https://privacy-scaling-explorations.github.io/sonobe-docs/design/nova-decider-onchain.html#the-decider-approach)) involves doing the random-linear-combination of scalars but also of the curve points which would be non-native in-circuit. Since they're points of the BN254 and have precompiles in the EVM we thought on computing them in Solidity directly as it's done.

Reply:
Ah I see, make sense!